pkg/storegateway/indexheader/reader_pool_test.go

```diff
@@ -138,3 +123,83 @@ func TestReaderPool_ShouldCloseIdleLazyReaders(t *testing.T) {
 	require.Equal(t, float64(2), promtestutil.ToFloat64(metrics.lazyReader.loadCount))
 	require.Equal(t, float64(2), promtestutil.ToFloat64(metrics.lazyReader.unloadCount))
 }
+
+func TestReaderPool_PersistLazyLoadedBlock(t *testing.T) {
+	const idleTimeout = time.Second
+	ctx, tmpDir, bkt, blockID, metrics := prepareReaderPool(t)
+
+	// Note that we are creating a ReaderPool that doesn't run a background cleanup task for idle
+	// Reader instances. We'll manually invoke the cleanup task when we need it as part of this test.
+	pool := newReaderPool(log.NewNopLogger(), true, idleTimeout, metrics)
+	defer pool.Close()
+
+	r, err := pool.NewBinaryReader(ctx, log.NewNopLogger(), bkt, tmpDir, blockID, 3, Config{})
+	require.NoError(t, err)
+	defer func() { require.NoError(t, r.Close()) }()
+
+	// Ensure it can read data.
+	labelNames, err := r.LabelNames()
+	require.NoError(t, err)
+	require.Equal(t, []string{"a"}, labelNames)
+	require.Equal(t, float64(1), promtestutil.ToFloat64(metrics.lazyReader.loadCount))
+	require.Equal(t, float64(0), promtestutil.ToFloat64(metrics.lazyReader.unloadCount))
+
+	snapshot := lazyLoadedHeadersSnapshot{
+		HeaderLastUsedTime: pool.LoadedBlocks(),
+		UserID:             "anonymous",
+	}
+
+	err = snapshot.persist(tmpDir)
+	require.NoError(t, err)
+	require.Greater(t, snapshot.HeaderLastUsedTime[blockID], int64(0), "lazyLoadedBlocks snapshot must be set")
+
+	persistedFile := filepath.Join(tmpDir, lazyLoadedHeadersListFile)
+	persistedData, err := os.ReadFile(persistedFile)
+	require.NoError(t, err)
+	expectedSnapshot := &lazyLoadedHeadersSnapshot{}
```

Comment:
The name `expectedSnapshot` is a bit misleading - this is the snapshot read from disk. Perhaps `snapshotReadFromDisk` would be clearer?

Reply:
> Why not extract it as part of this PR?

I was considering limiting the size of this PR because extracting the logic will touch the code of the shutdown marker, hence it will be better to have it as a separate PR.

Reply:
Eventually I extracted the fsync function in this PR.